src/cdn.rs

```diff
@@ -117,75 +258,202 @@ impl CdnBackend {
                     .build(),
             )
             .send()
-            .await?;
+            .await?
+            .invalidation()
+            .ok_or_else(|| {
+                anyhow!("missing invalidation information in create-invalidation result")
+            })?
+            .id()
+            .ok_or_else(|| anyhow!("missing invalidation ID in create-invalidation result"))?
+            .to_owned())
+    }
+}
 
-        Ok(())
+#[instrument(skip(conn))]
+pub(crate) fn handle_queued_invalidation_requests(
+    cdn: &CdnBackend,
+    conn: &mut impl postgres::GenericClient,
+    distribution_id: &str,
+) -> Result<()> {
+    info!("handling queued CDN invalidations");
+
+    let active_invalidations = cdn
+        .active_invalidations(distribution_id)
+        .context("error querying active invalidations")?;
+
+    // for now we assume all invalidation paths are wildcard invalidations,
+    // so we apply the wildcard limit.
+    let active_path_invalidations: usize = active_invalidations
+        .iter()
+        .map(|i| i.path_patterns.len())
+        .sum();
+
+    debug!(
+        active_invalidations = active_invalidations.len(),
+        active_path_invalidations, "found active invalidations",
+    );
+
+    // remove the invalidation from the queue when they are completed.
+    // We're only looking at InProgress invalidations,
+    // we don't differentiate between `Completed` ones, and invalidations
+    // missing in the CloudFront `ListInvalidations` response.
+    conn.execute(
+        "DELETE FROM cdn_invalidation_queue
+         WHERE 
+             cdn_distribution_id = $1 AND 
+             created_in_cdn IS NOT NULL AND 
+             NOT (cdn_reference = ANY($2))",
+        &[
+            &distribution_id,
+            &active_invalidations
+                .iter()
+                .map(|i| i.invalidation_id.clone())
+                .collect::<Vec<_>>(),
+        ],
+    )?;
+
+    let possible_path_invalidations: i32 =
+        MAX_CLOUDFRONT_WILDCARD_INVALIDATIONS - active_path_invalidations as i32;
```

Comment:
I'm a little confused by this. What did we conclude about the allowed number of paths vs the allowed number of invalidations? I know we can do 15 invalidations at once, but can we have 1 path per invalidation? 100? This code seems to imply each path counts against our limit of 15, which would eliminate one of the big wins for batching.

As I read it, this would create a single invalidation on the DB containing 13 paths, then would create no invalidations for a while (waiting for that one to finish), then it would create another invalidation on the DB containing 13 more paths. Is that correct?

Reply:
> I'm a little confused by this. What did we conclude about the allowed number of paths vs the allowed number of invalidations? I know we can do 15 invalidations at once, but can we have 1 path per invalidation? 100? This code seems to imply each path counts against our limit of 15, which would eliminate one of the big wins for batching.

I manually tested this. The limit is not per invocation, but for the whole distribution. 

You're right, there is no value in batching here, but in: 
- queueing -> so delaying invalidations if there are too many happening. 
- the potential for "escalating" them, for example: if invalidations are queued for >x hours, remove them, and invalidate the whole distribution. Or merging patterns. 

> 
> As I read it, this would create a single invalidation on the DB containing 13 paths, then would create no invalidations for a while (waiting for that one to finish), then it would create another invalidation on the DB containing 13 more paths. Is that correct?

Yep, assuming there are always >= 13 paths queued to invalidate.